src/branch.ts

```diff
@@ -0,0 +1,183 @@
+ *   pipe(data, branch((x) => x > 3, add(1), multiply(2)));
+ *   map(data, branch(isNullish, (x, index) => x + index));
+ * @dataLast
+ * @category Function
+ */
+export function branch<
+  T,
+  Args extends Array<any>,
+  Predicate extends (data: T, ...args: Args) => boolean,
+  OnTrue extends (data: GuardType<Predicate, T>, ...args: Args) => unknown,
+>(
+  predicate: Predicate,
+  onTrue: OnTrue,
+): (
+  data: T,
+  ...args: Args
+) => Exclude<T, GuardType<Predicate>> | ReturnType<OnTrue>;
+export function branch<
+  T,
+  Args extends Array<any>,
+  Predicate extends (data: T, ...args: Args) => boolean,
+  OnTrue extends (data: GuardType<Predicate, T>, ...args: Args) => unknown,
+  OnFalse extends (
+    data: Exclude<T, GuardType<Predicate>>,
+    ...args: Args
+  ) => unknown,
+>(
+  predicate: Predicate,
+  onTrue: OnTrue,
+  onFalse: OnFalse,
+): (data: T, ...args: Args) => ReturnType<OnFalse> | ReturnType<OnTrue>;
+
+/**
+ * Conditionally run a function based on a predicate, similar to
+ * the [ternary `?:`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_operator)
+ * operator. If the optional `onFalse` function is not provided, the data will
+ * be passed through.
+ *
+ * Supports type predicates to refine the types for both branches and the return
+ * value.
+ *
+ * Additional arguments are passed to all functions. In data-first calls, they
+ * are taken as variadic arguments; but in data-last calls, they are when the
+ * curried function itself is called.
+ *
+ * For more complex cases check out `conditional`.
+ *
+ * @param data - The data to be passed to all functions, as the first param.
+ * @param predicate - The function that decides which of the 2 branches would
+ * run. If a type-predicate is provided, then the types of both branches and the
+ * return value would be narrowed.
+ * @param onTrue - The function that would run when the predicate returns
+ * `true`.
+ * @signature
+ *   branch(data, predicate, onTrue)
+ *   branch(data, predicate, onTrue, onFalse, ...extraArgs)
+ * @example
+ *   branch(data, isNullish, constant(42));
+ *   branch(data, (x) => x > 3, add(1), multiply(2));
+ *   branch(data, isString, (x, radix) => parseInt(x, radix), identity(), 10);
+ * @dataLast
+ * @category Function
+ */
+export function branch<
+  T,
+  Predicate extends (data: T) => boolean,
+  OnTrue extends (data: GuardType<Predicate, T>) => unknown,
+>(
+  data: T,
+  predicate: Predicate,
+  onTrue: OnTrue,
+): Exclude<T, GuardType<Predicate>> | ReturnType<OnTrue>;
+export function branch<
+  T,
+  Args extends Array<any>,
+  Predicate extends (data: T, ...args: Args) => boolean,
+  OnTrue extends (data: GuardType<Predicate, T>, ...args: Args) => unknown,
+  OnFalse extends (
+    data: Exclude<T, GuardType<Predicate>>,
+    ...args: Args
+  ) => unknown,
+>(
+  data: T,
+  predicate: Predicate,
+  onTrue: OnTrue,
+  onFalse: OnFalse,
+  ...args: Args
```

Comment:
**thought:** the `...args` reading here is confusing to me, especially since you can only do it if you have the `onFalse` argument. i would be much happier if both `onTrue` and `onFalse` only took in `data`, or we only had the `dataLast` version and not the `dataFirst` version

Reply:
now that i think about it, the optional `onFalse` here is also weird. in particular it feels like this function is trying to do two different things, which should be two different functions. i think the ramda design of having a separate `defaultTo` and `ifElse` feels right to me

maybe we should split this into

```ts
// defaultValue, not a function
defaultIf(predicate, defaultValue)(data)

ifElse(predicate, onTrue, onFalse)(data)
```